core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java

```diff
@@ -0,0 +1,151 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryMetadataService {
+
+    public static String QUERY_METADATA = "_query_metadata_";
+    public static String QUERY_METADATA_FIELD_PREFIX = QUERY_METADATA + "field_";
+    public static String QUERY_METADATA_FIELD_UNKNOWN = QUERY_METADATA + "unknown";
+    public static FieldType QUERY_METADATA_FIELD_TYPE = new FieldType();
+
+    static {
+        QUERY_METADATA_FIELD_TYPE.setTokenized(false);
+        QUERY_METADATA_FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+        QUERY_METADATA_FIELD_TYPE.freeze();
+    }
+
+    public void extractQueryMetadata(Query query, ParseContext.Document document) {
+        List<Term> queryTerms = new ArrayList<>();
+        extractQueryMetadata(query, queryTerms);
+        for (Term term : queryTerms) {
+            document.add(new Field(term.field(), term.bytes(), QUERY_METADATA_FIELD_TYPE));
+        }
+    }
+
+    public void extractQueryMetadata(Query query, List<Term> queryTerms) {
+        if (query instanceof TermQuery) {
+            Term term = ((TermQuery) query).getTerm();
+            queryTerms.add(new Term(QUERY_METADATA_FIELD_PREFIX + term.field(), term.bytes()));
+        } else if (query instanceof PhraseQuery) {
+            Term[] terms = ((PhraseQuery) query).getTerms();
+            if (terms.length == 0) {
+                return;
+            }
+
+            Term longestTerm = terms[0];
+            for (Term term : terms) {
+                if (longestTerm.bytes().length < term.bytes().length) {
+                    longestTerm = term;
+                }
+            }
+            queryTerms.add(new Term(QUERY_METADATA_FIELD_PREFIX + longestTerm.field(), longestTerm.bytes()));
+        } else if (query instanceof BooleanQuery) {
+            List<BooleanClause> clauses = ((BooleanQuery) query).clauses();
+            boolean noShouldClauses = true;
+            for (BooleanClause clause : clauses) {
+                if (clause.getOccur() == BooleanClause.Occur.SHOULD) {
+                    noShouldClauses = false;
+                    break;
+                }
+            }
+            if (noShouldClauses) {
+                List<Term> bestClause = null;
+                for (BooleanClause clause : clauses) {
+                    if (clause.isProhibited()) {
+                        // we don't need to remember the things that do *not* match...
+                        continue;
+                    }
+
+                    List<Term> temp = new ArrayList<>();
+                    extractQueryMetadata(clause.getQuery(), temp);
+                    if (bestClause == null) {
+                        bestClause = temp;
+                    } else {
+                        int currentSize = 0;
+                        for (Term term : bestClause) {
+                            currentSize += term.bytes().length;
+                        }
```

Comment:
can you make it a utility method to make the code a bit easier to read? (compute the sum of term lengths)